src/dispense_tracker.rs

```diff
@@ -0,0 +1,96 @@
+use std::{
+    cmp::Ordering,
+    collections::{BinaryHeap, HashMap, HashSet},
+};
+
+use fuel_types::Address;
+
+#[derive(Debug, Eq, PartialEq)]
+pub(crate) struct Entry {
+    address: Address,
+    timestamp: u64,
+}
+
+impl Ord for Entry {
+    fn cmp(&self, other: &Self) -> Ordering {
+        other.timestamp.cmp(&self.timestamp)
+    }
+}
+
+impl PartialOrd for Entry {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+pub trait Clock: std::fmt::Debug + Send + Sync {
+    fn now(&self) -> u64;
+}
+
+#[derive(Debug)]
+pub struct TokioTime {}
+
+impl Clock for TokioTime {
+    fn now(&self) -> u64 {
+        tokio::time::Instant::now().elapsed().as_secs()
+    }
+}
+
+#[derive(Debug)]
+pub struct DispenseTracker {
+    tracked: HashMap<Address, u64>,
```

Comment:
It looks like we keep the tracking in application memory. This makes sense as long as there is only one instance of the faucet and that single instance remains up! But if we ever needed to restart the server or introduce load balancing, we might want a more sophisticated solution like a Redis cache in a separate service.

Reply:
It's okay to store it in the memory; even with crashes or restarts, it still prevents users from spamming us. If they can't force the crash of the faucet, then it is okay=)